dev/plugins/hu.elte.txtuml.diagnostics/src/hu/elte/txtuml/diagnostics/animation/js/DiagnosticsServer.java

```diff
@@ -0,0 +1,127 @@
+package hu.elte.txtuml.diagnostics.animation.js;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+import com.sun.net.httpserver.Headers;
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+
+import hu.elte.txtuml.api.model.execution.diagnostics.protocol.GlobalSettings;
+import hu.elte.txtuml.api.model.execution.diagnostics.protocol.MessageType;
+import hu.elte.txtuml.api.model.execution.diagnostics.protocol.ModelEvent;
+import hu.elte.txtuml.diagnostics.session.DiagnosticsPlugin;
+
+/**
+ * Serves diagnostics data over HTTP.
+ */
+public class DiagnosticsServer {
+
+	private HttpServer server;
+	private ConcurrentMap<String, RegistryEntry> registry = new ConcurrentHashMap<>();
+	private DiagnosticsPlugin diagnosticsPlugin;
+	
+	public DiagnosticsServer(DiagnosticsPlugin diagnosticsPlugin){
+		this.diagnosticsPlugin = diagnosticsPlugin;
+	}
+
+	public void start(Integer port) throws IOException {
+		server = HttpServer.create(new InetSocketAddress(port), 0);
+		server.createContext("/" + GlobalSettings.TXTUML_DIAGNOSTICS_HTTP_PATH, new DiagnosticsHandler());
+		server.createContext("/delay", new DelayHandler());
+		server.setExecutor(null); // creates a default executor
+		server.start();
+	}
+
+	public void stop() {
+		if (server != null) server.stop(0);
+		registry.clear();
+	}
+
+	public void register(ModelEvent event) {
+		RegistryEntry registryEntry = new RegistryEntry(event.modelClassName,
+				event.modelClassInstanceName, event.eventTargetClassName);
+		registry.put(event.modelClassInstanceID, registryEntry);
+	}
+
+	private class DiagnosticsHandler implements HttpHandler {
+
+		@Override
+		public void handle(HttpExchange exchange) throws IOException {
+			// Set required headers
+			Headers headers = exchange.getResponseHeaders();
+			headers.add("Content-type", "application/json");
+			headers.add("Access-Control-Allow-Origin", "*");
+
+			// Build the payload
+			String response = registry.entrySet().stream()
+					.map(DiagnosticsServer::registryEntryToJson)
+					.collect(Collectors.joining(",", "[", "]"));
+
+			// Write response
+			exchange.sendResponseHeaders(200, response.length());
+			OutputStream os = exchange.getResponseBody();
+			os.write(response.getBytes());
+			os.close();
+		}
+
+	}
+	
+	private class DelayHandler implements HttpHandler {
+
+		@Override
+		public void handle(HttpExchange exchange) throws IOException {
+			//Get request data
+			StringBuilder body = new StringBuilder();
+		    try (InputStreamReader reader = new InputStreamReader(exchange.getRequestBody())) {
+		        char[] buffer = new char[256];
```

Comment:
Is this just an `InputStream` to `String` conversion? If yes, I recommend to use something more intuitive, e.g. `String body = new BufferedReader(reader).lines().collect(Collectors.joining("\n"))`.